main.py

```diff
@@ -41,10 +37,25 @@ def produce_task(logger, session, server, task):
 
         # pull out the results the caller requested, ignore any others that were exposed
         output_keys = json.loads(task.output_keys)
+
+        # loop over the (ordered) list of requested output types until we find one that we support
+        output_types = json.loads(task.output_types)
+        selected_output_type = None
+        for output_type in output_types:
+            if output_type in messages.ALLOWED_TYPES:
+                selected_output_type = output_type
+                break
+        if not selected_output_type:
+            raise ValueError(f"could not write results in requested formats '{output_types}'")
+
         for output_key in output_keys:
             if output_key in results.values:
-                preds_path = utils.make_preds_filename(task.request_id, output_key=output_key)
-                results.values[output_key].to_csv(preds_path, index=False)
+                if selected_output_type == ValueType.PARQUET_URI:
+                    preds_path = utils.make_preds_filename(task.request_id, output_key=output_key, output_type=selected_output_type)
+                    results.values[output_key].to_parquet(preds_path, index=False)
+                elif selected_output_type == ValueType.CSV_URI:
+                    preds_path = utils.make_preds_filename(task.request_id, output_key=output_key, output_type=selected_output_type)
```

Comment:
This line could be outside the ifs I think

Reply:
Needs the current `output_key` value when generating the path.